### PR TITLE
Add blog post about stream consumer performance

### PR DESCRIPTION
**Pull request by rrhodes**
New blog post about improving AWS Stream Consumer service performance. Raised an initial draft of this content in #tech-guilty-publicity for review and feedback - happy to address additional feedback here before publication.

### CODE REVIEW

**Review by buck06191 — COMMENTED** [review]
Two main things:

All image tags should have alt text for people using screen readers. 

It would be nice for numbered lists to use consecutive numbers even in markdown

File: _posts/2024-05-13-stream-consumer-performance.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+To capture and fan-out dispatch changes from DynamoDB, we leveraged [DynamoDB Streams](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Streams.html): a stream integration managed by AWS which records data inserted, updated, and deleted from the DynamoDB table. Services that must process dispatch changes may synchronously consume stream records, analyze what data has changed, and act accordingly on those changes. For Power API, four services were set up to consume DynamoDB Stream records: one [EventBridge Pipe](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-pipes.html) - explored in [a previous Kraken Tech blog post](https://tech.octopus.energy/news/2024/02/19/using-eventbridge-pipes.html) - and three [Lambda functions](https://docs.aws.amazon.com/lambda/latest/dg/welcome.html).
+
+<img src="/assets/img/posts/2024-05-13-stream-consumer-performance/dynamodb-stream-consumers.png"/>
```
[the remark this answers is not in the document]
Comment (rrhodes) [reply _posts/2024-05-13-stream-consumer-performance.md]: Done: [bee366c](https://github.com/octoenergy/octoenergy.github.io/pull/69/commits/bee366cfd258a0905f0491943fb14da853d486b6)

File: _posts/2024-05-13-stream-consumer-performance.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+<img src="/assets/img/posts/2024-05-13-stream-consumer-performance/dynamodb-iterator-age-maximum.png"/>
+
+Based on these findings, we explored options to lower the iterator age and thus decrease the time it takes for services to successfully process records from the stream. In particular, we wanted to reduce the frequent spikes in _average_ and _maximum_ iterator ages observed.
+
+## Debugging High Consumer Iterator Ages
+
+There were [a few avenues](https://repost.aws/knowledge-center/dynamodb-lambda-iteratorage) to explore which may explain high iterator ages:
+
+1. A consumer fails to process a record. In which case, the record is returned to the stream for the consumer to retry. Multiple executions of the consumer will increase the iterator age.
+1. Low consumer throughput. If a consumer processes records at a slower rate than they are written, then a backlog of records builds up in the stream, which adds to the iterator age.
```
[the remark this answers is not in the document]
Comment (rrhodes) [reply _posts/2024-05-13-stream-consumer-performance.md]: Done: [bee366c](https://github.com/octoenergy/octoenergy.github.io/pull/69/commits/bee366cfd258a0905f0491943fb14da853d486b6)

File: _posts/2024-05-13-stream-consumer-performance.md [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+<img src="/assets/img/posts/2024-05-13-stream-consumer-performance/kinesis-iterator-age-average.png"/>
+
+Furthermore, the maximum iterator age metric did cap at 30s, an improvement from the previously observed maximum of 40s. Iterator ages upward of 15s occurred rarely. Where higher ages occur, those are likely attributable to spikes in traffic or Lambda consumer timeouts.
+
+<img src="/assets/img/posts/2024-05-13-stream-consumer-performance/kinesis-iterator-age-maximum.png"/>
+
+The general drop in the iterator age was a significant win for the team. In summary, three changes of increasing complexity led to this improvement:
+
+1. Optimizing the Batch Size and Parallelization Factor of stream integrations.
+1. Improving performance of stream consumer dependencies (e.g. APIs) for faster processing.
```
[the remark this answers is not in the document]
Comment (rrhodes) [reply _posts/2024-05-13-stream-consumer-performance.md]: Done: [bee366c](https://github.com/octoenergy/octoenergy.github.io/pull/69/commits/bee366cfd258a0905f0491943fb14da853d486b6)

File: _posts/2024-05-13-stream-consumer-performance.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
+---
+title: Optimizing AWS Stream Consumer Performance
+layout: post
+category: news
+author: Ross Rhodes
+banner: /assets/img/posts/2015-11-23-tech-jobs.jpg
+hex: 0e1720
+---
+
```
[the remark this answers is not in the document]
Comment (rrhodes) [reply _posts/2024-05-13-stream-consumer-performance.md:9]: Good shout. I've added an intro now: [24331d9](https://github.com/octoenergy/octoenergy.github.io/pull/69/commits/24331d96e1a26897b18f79eee543969f15ce6bd6)

Let me know what you think!

**Review by frederizia — APPROVED** [review]
I really like how you're talking through all the options and give us an insight into the problem solving process.

Out of interest, how long did all this take? Investigation, introducing the fan-out function and then the migration?

**Comment by rrhodes** [comment]
> Out of interest, how long did all this take? Investigation, introducing the fan-out function and then the migration?

Good question! For a couple engineers, this story spans a few weeks. The initial investigation didn't take too long, most of our time was dedicated to the migration step with Kinesis, since this involved new metrics, alarms, infrastructure, etc.

File: _posts/2024-05-13-stream-consumer-performance.md [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+<img src="/assets/img/posts/2024-05-13-stream-consumer-performance/dynamodb-iterator-age-maximum.png" alt="DynamoDB Stream Maximum Iterator Age"/>
+
+Based on these findings, we explored options to lower the iterator age and thus decrease the time it takes for services to successfully process records from the stream. In particular, we wanted to reduce the frequent spikes in _average_ and _maximum_ iterator ages observed.
+
+## Debugging High Consumer Iterator Ages
+
+There were [a few avenues](https://repost.aws/knowledge-center/dynamodb-lambda-iteratorage) to explore which may explain high iterator ages:
+
+1. A consumer fails to process a record. In which case, the record is returned to the stream for the consumer to retry. Multiple executions of the consumer will increase the iterator age.
```
[the remark this answers is not in the document]
Comment (rrhodes) [reply _posts/2024-05-13-stream-consumer-performance.md]: Good shout! I've swapped out `which` for `this`.

**Review by buck06191 — APPROVED** [review]

**Pull request merged**